packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol

```diff
@@ -0,0 +1,285 @@
+
+        emit SeededWindow(windowIndex, totalRewardsDistributed, windowStart, windowEnd, rewardToken);
+    }
+
+    function _claimWindow(Claim memory claim) internal windowNotLocked(claim.windowIndex) {
+        // Check claimed proof against merkle window at given index.
+        require(verifyClaim(claim), "Incorrect merkle proof");
+        // Check the account has not yet claimed their full amount for this window.
+        require(
+            amountClaimedFromWindow[claim.windowIndex][claim.account] < claim.amount,
+            "Account has already claimed the full amount for this window"
+        );
+
+        // Proof is correct; check that claimant has enough vested amount to disburse and decrease their balance.
+        Window memory merkleWindow = merkleWindows[claim.windowIndex];
+        (uint256 amountVestedNetPreviousClaims, uint256 amountRemaining) = _debitClaimedAmount(claim, merkleWindow);
+        require(amountVestedNetPreviousClaims > 0, "Zero vested amount");
+
+        // `_debitClaimedAmount` would have reverted if `amountVestedNetPreviousClaims == 0` so we can
+        // assume that we're disbursing a non-0 amount of reward tokens.
+        _disburse(claim, merkleWindow, amountVestedNetPreviousClaims);
+
+        emit Claimed(
+            msg.sender,
+            claim.windowIndex,
+            claim.account,
+            claim.amount,
+            amountVestedNetPreviousClaims,
+            amountRemaining,
+            address(merkleWindow.rewardToken)
+        );
+    }
+
+    function _debitClaimedAmount(Claim memory claim, Window memory merkleWindow)
+        internal
+        returns (uint256 amountVestedNetPreviousClaims, uint256 amountRemaining)
+    {
+        uint256 currentTime = getCurrentTime();
+
+        // Calculate how much of the claimed amount has vested. This will return 0 if the contract time is less than or
+        // equal to the window start, or return the full `amount` if the contract time is greater than or equal to the window end.
+        uint256 amountVested = calcVestedAmount(claim.amount, currentTime, merkleWindow.start, merkleWindow.end);
+
+        // Calculate how much the account has left to claim (unvested). Useful in logging.
+        // This is the maximum claim reward for this function call.
+        amountRemaining = claim.amount.sub(amountVested);
+
+        // Fetch how much the account has previously claimed for this window.
+        uint256 amountPreviouslyClaimed = amountClaimedFromWindow[claim.windowIndex][claim.account];
+
+        // The account can claim their maximum vested amount minus any previous claim amounts.
+        amountVestedNetPreviousClaims = amountVested.sub(amountPreviouslyClaimed);
+
+        // The user has now claimed their full vested amount this far into the window.
+        // Update their previously claimed amount.
+        amountClaimedFromWindow[claim.windowIndex][claim.account] = amountVested;
```

Comment:
this is being set with the amountVested, which means its possible to have additional unclaimed funds if you claimed before expiry? This would make more sense to me if it was set to claim.amount instead, but I could be wrong on this

Reply:
So imagine that `claim.amount > amountVested` (because you're in the middle of the vesting period), then the `amountClaimedFromWindow` should be set to the total amount vested. The assumption we're making is that every call to `claimWindow` claims ALL of your vested tokens, so the `amountClaimedFromWindow` is always set to the maximum vested amount when you call claim.

Does that make sense?

Reply:
I see now, yea so you never actually lose rewards by claiming early. This makes a lot of sense and most my comments are around this confusion.

Reply:
Yeah basically you just call `claim()` with some proof parameters, and if the proof succeeds, you get back the maximum that you are owed.